File: frontend/types/relationship.ts
/**
 * Relationship TypeScript types — mirrors backend relationship schemas.
 */

// ── Enums ──────────────────────────────────────────────────────────────────────

export type EntityType =
  | "character"
  | "location"
  | "organization"
  | "object"
  | "world_rule";

export type RelationshipDirection = "unidirectional" | "bidirectional";

export type RelationshipType =
  | "ally_of"
  | "enemy_of"
  | "friend_of"
  | "parent_of"
  | "child_of"
  | "sibling_of"
  | "mentor_of"
  | "student_of"
  | "member_of"
  | "leader_of"
  | "owns"
  | "created"
  | "created_by"
  | "located_in"
  | "lives_in"
  | "protects"
  | "rules"
  | "loves"
  | "hates"
  | "rival_of"
  | "custom";

// ── Model ──────────────────────────────────────────────────────────────────────

export interface Relationship {
  id: string;
  universe_id: string;
  source_entity_type: EntityType;
  source_entity_id: string;
  target_entity_type: EntityType;
  target_entity_id: string;
  relationship_type: RelationshipType;
  title: string | null;
  description: string | null;
  strength: number | null;
  direction: RelationshipDirection;
  metadata: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface RelationshipList {
  items: Relationship[];
  total: number;
  limit: number;
  offset: number;
}

// ── Request payloads ───────────────────────────────────────────────────────────

export interface CreateRelationshipRequest {
  universe_id: string;
  source_entity_type: EntityType;
  source_entity_id: string;
  target_entity_type: EntityType;
  target_entity_id: string;
  relationship_type: RelationshipType;
  title?: string;
  description?: string;
  strength?: number;
  direction?: RelationshipDirection;
  metadata?: Record<string, unknown>;
}

export interface UpdateRelationshipRequest {
  relationship_type?: RelationshipType;
  title?: string;
  description?: string;
  strength?: number;
  direction?: RelationshipDirection;
  metadata?: Record<string, unknown>;
}

// ── Display helpers ────────────────────────────────────────────────────────────

export const RELATIONSHIP_TYPE_LABELS: Record<RelationshipType, string> = {
  ally_of: "Ally Of",
  enemy_of: "Enemy Of",
  friend_of: "Friend Of",
  parent_of: "Parent Of",
  child_of: "Child Of",
  sibling_of: "Sibling Of",
  mentor_of: "Mentor Of",
  student_of: "Student Of",
  member_of: "Member Of",
  leader_of: "Leader Of",
  owns: "Owns",
  created: "Created",
  created_by: "Created By",
  located_in: "Located In",
  lives_in: "Lives In",
  protects: "Protects",
  rules: "Rules",
  loves: "Loves",
  hates: "Hates",
  rival_of: "Rival Of",
  custom: "Custom",
};

export const ENTITY_TYPE_LABELS: Record<EntityType, string> = {
  character: "Character",
  location: "Location",
  organization: "Organization",
  object: "Object",
  world_rule: "World Rule",
};

export const DIRECTION_LABELS: Record<RelationshipDirection, string> = {
  unidirectional: "One-way →",
  bidirectional: "Two-way ↔",
};

File: backend/app/models/relationship.py
from enum import StrEnum

from sqlalchemy import ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base
from app.database.mixins import BaseEntity


class EntityType(StrEnum):
    CHARACTER = "character"
    LOCATION = "location"
    ORGANIZATION = "organization"
    WORLD_OBJECT = "world_object"
    WORLD_RULE = "world_rule"
    character = "character"
    location = "location"
    organization = "organization"
    world_object = "world_object"
    world_rule = "world_rule"


class RelationshipType(StrEnum):
    ALLY_OF = "ally_of"
    ENEMY_OF = "enemy_of"
    LIVES_IN = "lives_in"
    MEMBER_OF = "member_of"
    ally_of = "ally_of"
    enemy_of = "enemy_of"
    lives_in = "lives_in"
    member_of = "member_of"


class RelationshipDirection(StrEnum):
    UNIDIRECTIONAL = "unidirectional"
    BIDIRECTIONAL = "bidirectional"
    unidirectional = "unidirectional"
    bidirectional = "bidirectional"


class Relationship(BaseEntity, Base):
    __tablename__ = "relationships"
    universe_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("universes.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    source_entity_type: Mapped[str] = mapped_column(String(50), nullable=False)
    source_entity_id: Mapped[str] = mapped_column(
        String(36), nullable=False, index=True
    )
    target_entity_type: Mapped[str] = mapped_column(String(50), nullable=False)
    target_entity_id: Mapped[str] = mapped_column(
        String(36), nullable=False, index=True
    )
    relationship_type: Mapped[str] = mapped_column(
        String(50), nullable=False, index=True
    )
    title: Mapped[str | None] = mapped_column(String(300), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    strength: Mapped[int | None] = mapped_column(Integer, nullable=True)
    direction: Mapped[str] = mapped_column(
        String(20), nullable=False, server_default="unidirectional"
    )
    metadata_json: Mapped[str | None] = mapped_column(Text, nullable=True)
